index.js

```diff
@@ -16,114 +16,114 @@ function helmetPlugin (fastify, options, next) {
     fastify.decorateReply('helmet', null)
   }
 
-  // We will add the onRequest helmet middleware functions through the onRoute hook if needed
+  // We initialize the `cspNonce` reply decorator only if it does not already exists
+  if (!fastify.hasReplyDecorator('cspNonce')) {
+    fastify.decorateReply('cspNonce', null)
+  }
+
   fastify.addHook('onRoute', (routeOptions) => {
     if (typeof routeOptions.helmet !== 'undefined') {
       if (typeof routeOptions.helmet === 'object') {
-        const { enableCSPNonces: enableRouteCSPNonces, ...helmetRouteConfiguration } = routeOptions.helmet
-
-        // If route helmet options are set they overwrite the global helmet configuration
-        const mergedHelmetConfiguration = Object.assign({}, globalConfiguration, helmetRouteConfiguration)
-
-        buildRouteHooks(mergedHelmetConfiguration, routeOptions)
-
-        if (enableRouteCSPNonces) {
-          routeOptions.onRequest.push(buildCSPNonce(fastify, mergedHelmetConfiguration))
-        }
+        routeOptions.config = Object.assign(routeOptions.config || Object.create(null), { helmet: routeOptions.helmet })
       } else if (routeOptions.helmet === false) {
-        // don't apply any helmet settings but decorate the reply with a fallback to the
-        // global helmet options
-        buildRouteHooks(globalConfiguration, routeOptions, true)
+        routeOptions.config = Object.assign(routeOptions.config || Object.create(null), { helmet: { skipRoute: true } })
       } else {
         throw new Error('Unknown value for route helmet configuration')
       }
-    } else if (isGlobal) {
-      // if the plugin is set globally (meaning that all the routes will be decorated)
-      // As the endpoint, does not have a custom helmet configuration, use the global one.
-      buildRouteHooks(globalConfiguration, routeOptions)
+    }
+  })
 
-      if (enableCSPNonces) {
-        routeOptions.onRequest.push(buildCSPNonce(fastify, globalConfiguration))
-      }
+  fastify.addHook('onRequest', async (request, reply) => {
+    const routeOptions = request.context.config && typeof request.context.config.helmet !== 'undefined'
```

Comment:
`request.context.config` should be always exist.

https://github.com/fastify/fastify/blob/70749bcc278ae9c4d1487ffc1bdecd67fb8bc14e/lib/fourOhFour.js#L127
https://github.com/fastify/fastify/blob/70749bcc278ae9c4d1487ffc1bdecd67fb8bc14e/lib/route.js#L226-L230

Reply:
done and I added a few more comments to the code. ^^